File: legacy/capacitor-web/src/services/emotionalAnalysisService.js
class EmotionalAnalysisService {
  constructor() {
    this.apiKey = process.env.REACT_APP_OPENAI_API_KEY || '';
    this.baseURL = 'https://api.openai.com/v1';
    this.modelName = 'gpt-4o'; // Using OpenAI GPT-4o model
  }

  async analyzeEmotionalScores(messages) {
    console.log('🧠 Starting emotional analysis...');
    console.log('🔍 EMOTIONAL DEBUG: messages type:', typeof messages, 'length:', messages?.length);
    
    try {
      // Extract conversation text from messages
      const conversationText = messages
        .filter(msg => msg.text && msg.text.trim())
        .map(msg => `${msg.sender === 'user' ? 'User' : 'Assistant'}: ${msg.text}`)
        .join('\n');
      
      if (!conversationText || conversationText.trim() === '') {
        console.log('⚠️ No conversation text found, returning default scores');
        return this.getDefaultScores();
      }
      
      // Create emotional analysis prompt
      const prompt = `Analyze the emotional state from this conversation and provide numerical scores (0-100) for:
- happiness (how positive/joyful)
- energy (how energetic/motivated)  
- anxiety (how worried/anxious)
- stress (how stressed/pressured)

Conversation:
${conversationText}

Respond ONLY with a JSON object in this exact format:
{
  "happiness": <number>,
  "energy": <number>,
  "anxiety": <number>,
  "stress": <number>
}`;

      // Use OpenAI API
      const apiUrl = `${this.baseURL}/chat/completions`;
      
        try {
        console.log('🤖 Using OpenAI for emotional analysis');
        console.log('🌐 API URL:', apiUrl);
          
          // Create AbortController for timeout
          const controller = new AbortController();
          const timeoutId = setTimeout(() => controller.abort(), 120000); // 120 seconds
          
          const response = await fetch(apiUrl, {
            method: 'POST',
            headers: {
              'Content-Type': 'application/json',
              'Authorization': `Bearer ${this.apiKey}`
            },
            body: JSON.stringify({
              model: this.modelName,
              messages: [{
                role: 'user',
                content: prompt
              }],
              temperature: 0.3,
              max_tokens: 300
            }),
            signal: controller.signal
          });
          
          clearTimeout(timeoutId);
          
          if (!response.ok) {
          const errorText = await response.text();
          console.error(`❌ OpenAI API failed:`, response.status, errorText);
          throw new Error(`Emotional analysis failed: ${response.status} ${response.statusText}`);
          }
          
          const data = await response.json();
          console.log('✅ EMOTIONAL DEBUG: Received response:', data);
          
          // Parse OpenAI API response format
          let responseText = '';
          if (data.choices && data.choices[0] && data.choices[0].message && data.choices[0].message.content) {
            responseText = data.choices[0].message.content;
          }
          
          // Extract JSON from response
          const jsonMatch = responseText.match(/\{[\s\S]*\}/);
          if (jsonMatch) {
            const emotionalData = JSON.parse(jsonMatch[0]);
            
            // Validate the scores
            if (this.isValidAnalysisResult(emotionalData)) {
              console.log('✅ Valid emotional analysis:', emotionalData);
              return emotionalData;
            } else {
              console.log('⚠️ Invalid emotional analysis format, using defaults');
              return this.getDefaultScores();
            }
          } else {
            console.log('⚠️ Could not extract JSON from response');
            return this.getDefaultScores();
          }
      } catch (error) {
        console.error(`❌ Error with OpenAI API:`, error);
      return this.getDefaultScores();
      }
    } catch (error) {
      console.error('❌ EMOTIONAL DEBUG: Error in analyzeEmotionalScores:', error);
      return this.getDefaultScores();
    }
  }

  isValidAnalysisResult(result) {
    if (!result || typeof result !== 'object') {
      return false;
    }
    
    const requiredFields = ['happiness', 'energy', 'anxiety', 'stress'];
    for (const field of requiredFields) {
      if (typeof result[field] !== 'number' || result[field] < 1 || result[field] > 100) {
        return false;
      }
    }
    
    return true;
  }

  getDefaultScores() {
    return {
      happiness: 50,
      energy: 50,
      anxiety: 30,
      stress: 30
    };
  }

  async saveEmotionalData(userId, dateId, scores) {
    // This method is called but not needed in the new implementation
    // The scores are already saved by ChatPage via firestoreService
    console.log('💾 Save emotional data called for:', userId, dateId, scores);
    return { success: true };
  }
}

export default new EmotionalAnalysisService();

File: flutter_app/lib/services/emotional_analysis_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config/env.dart';
import '../models/chat_message.dart';
import 'vertex_api_client.dart';

/// Mirrors src/services/emotionalAnalysisService.js
class EmotionalAnalysisService {
  EmotionalAnalysisService._();
  static final EmotionalAnalysisService instance = EmotionalAnalysisService._();

  Future<Map<String, int>> analyzeEmotionalScores(List<ChatMessage> messages) async {
    try {
      final conversationText = messages
          .where((m) => m.text.trim().isNotEmpty)
          .map((m) => '${m.sender == 'user' ? 'User' : 'Assistant'}: ${m.text}')
          .join('\n');
      if (conversationText.trim().isEmpty) return getDefaultScores();

      final prompt = '''Analyze the emotional state from this conversation and provide numerical scores (0-100) for:
- happiness
- energy
- anxiety
- stress

Conversation:
$conversationText

Respond ONLY with JSON: {"happiness":N,"energy":N,"anxiety":N,"stress":N}''';

      String responseText;
      if (isVertexBackendConfigured()) {
        responseText = await vertexGenerateContent(prompt: prompt, temperature: 0.3, maxOutputTokens: 300);
      } else {
        responseText = await _openAiComplete(prompt);
      }

      final match = RegExp(r'\{[\s\S]*\}').firstMatch(responseText);
      if (match != null) {
        final data = jsonDecode(match.group(0)!) as Map<String, dynamic>;
        if (_isValidAnalysisResult(data)) {
          return {
            'happiness': (data['happiness'] as num).round(),
            'energy': (data['energy'] as num).round(),
            'anxiety': (data['anxiety'] as num).round(),
            'stress': (data['stress'] as num).round(),
          };
        }
      }
      return getDefaultScores();
    } catch (_) {
      return getDefaultScores();
    }
  }

  Future<String> _openAiComplete(String prompt) async {
    final key = Env.openAiApiKey.trim();
    if (key.isEmpty) throw Exception('OpenAI key not configured');
    final res = await http.post(
      Uri.parse('https://api.openai.com/v1/chat/completions'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $key',
      },
      body: jsonEncode({
        'model': 'gpt-4o',
        'messages': [
          {'role': 'user', 'content': prompt},
        ],
        'temperature': 0.3,
        'max_tokens': 300,
      }),
    ).timeout(const Duration(seconds: 120));
    final data = jsonDecode(res.body) as Map<String, dynamic>;
    return data['choices']?[0]?['message']?['content'] as String? ?? '';
  }

  bool _isValidAnalysisResult(Map<String, dynamic> result) {
    for (final field in ['happiness', 'energy', 'anxiety', 'stress']) {
      final v = result[field];
      if (v is! num || v < 1 || v > 100) return false;
    }
    return true;
  }

  Map<String, int> getDefaultScores() => {
        'happiness': 50,
        'energy': 50,
        'anxiety': 30,
        'stress': 30,
      };
}

final emotionalAnalysisService = EmotionalAnalysisService.instance;
